### PR TITLE
GH-3594 Only do toString on the Number value when getLabel is called

### PR DESCRIPTION
**Pull request by JervenBolleman**
Signed-off-by: Jerven Bolleman <jerven.bolleman@sib.swiss>

GitHub issue resolved: #3594 

Briefly describe the changes proposed in this PR:

For a NumericLiteral store only the Number value and datatype.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java
Comment (hmottestad) [inline core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java:44]: This nullifies the sorting performance improvements that depend on the XSD.Datatype enum. If we end up merging https://github.com/eclipse/rdf4j/pull/3596 then it won't matter, but if not it would be worthwhile to implement support for XSD.Datatype here in this class.

**Review by hmottestad — CHANGES_REQUESTED** [review]
Can you hold on from merging this until we've had a discussion about my proposal for a new datatype enum: https://github.com/eclipse/rdf4j/pull/3596

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java
Comment (JervenBolleman) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java:44]: I would like to merge the XSD.Datatype work into this when it is ready.

**Comment by hmottestad** [comment]
@JervenBolleman I've merged the CoreDatatype PR. If you rebase your branch I can double check it for you before you merge it.

**Comment by JervenBolleman** [comment]
@hmottestad this has been rebased with the coredatype work used.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java
Comment (hmottestad) [inline core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java:156]: I cleaned up the equals method a bit. Since datatype is an enum it is faster to check than the number value, that's why I switched them around.

**Review by hmottestad — APPROVED** [review]
I've considered if this would have an impact for our users (and those who extend RDF4J) wrt. that I want to keep the changes between M3 and the final 4.0.0 release as small as possible. I think it should be fine.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java
Comment (JervenBolleman) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java:156]: Good for me.

**Comment by JervenBolleman** [comment]
I will wait for the checks to run and then merge and close the issue.

**Pull request merged**
